File: java/leetcode_grind/Day703.java
package leetcode_grind;

public class Day703 {
    // https://leetcode.com/problems/find-kth-bit-in-nth-binary-string/description/
    static class Solution1 {
        public char findKthBit(int n, int k) {
            int invertCount = 0;
            int len = (1 << n) - 1;
            while (k > 1) {
                if (k == len / 2 + 1) {
                    return invertCount % 2 == 0 ? '1' : '0';
                }

                if (k > len / 2) {
                    k = len + 1 - k;
                    invertCount++;
                }

                len /= 2;
            }
            return invertCount % 2 == 0 ? '0' : '1';
        }
    }
}

File: src/leetcode_grind/day_703.rs
// https://leetcode.com/problems/find-kth-bit-in-nth-binary-string/
pub fn find_kth_bit(n: i32, k: i32) -> char {
    let mut pos_in_section = k & -k;
    let mut is_in_inverted_part = (((k / pos_in_section) >> 1) & 1) == 1;
    let mut original_bit_is_one = (k & 1) == 0;
    if is_in_inverted_part {
        if original_bit_is_one {
            '0'
        } else {
            '1'
        }
    } else {
        if original_bit_is_one {
            '1'
        } else {
            '0'
        }
    }
}
